### PR TITLE
Add support for certificate policy qualifiers.

### LINKED ISSUE #230: Support for Policy Qualifiers
Let's Encrypt needs to set the PKIX CPS Pointer Qualifier in the Certificate template on our EE certs, matching our [recently-generated Intermediate (asn1 viewer)](http://lapo.it/asn1js/#3082058D30820375A003020102021100E79390BE92070349185F797581E5CA83300D06092A864886F70D01010B0500304F310B300906035504061302555331293027060355040A1320496E7465726E65742053656375726974792052657365617263682047726F7570311530130603550403130C4953524720526F6F74205831301E170D3135303630343132303032305A170D3230303630343132303032305A304A310B300906035504061302555331163014060355040A130D4C6574277320456E6372797074312330210603550403131A4C6574277320456E637279707420417574686F7269747920583130820122300D06092A864886F70D01010105000382010F003082010A02820101009CD30CF05AE52E47B7725D3783B3686330EAD735261925E1BDBE35F170922FB7B84B4105ABA99E350858ECB12AC468870BA3E375E4E6F3A76271BA7981601FD7919A9FF3D0786771C8690E9591CFFEE699E9603C48CC7ECA4D7712249D471B5AEBB9EC1E37001C9CAC7BA705EACE4AEBBD41E53698B9CBFD6D3C9668DF232A42900C867467C87FA59AB8526114133F65E98287CBDBFA0E56F68689F3853F9786AFB0DC1AEF6B0D95167DC42BA065B299043675806BAC4AF31B9049782FA2964F2A20252904C674C0D031CD8F31389516BAA833B843F1B11FC3307FA27931133D2D36F8E3FCF2336AB93931C5AFC48D0D1D641633AAFA8429B6D40BC0D87DC3930203010001A382016730820163300E0603551D0F0101FF04040302018630120603551D130101FF040830060101FF02010030540603551D20044D304B3008060667810C010201303F060B2B0601040182DF130101013030302E06082B060105050702011622687474703A2F2F6370732E726F6F742D78312E6C657473656E63727970742E6F7267301D0603551D0E04160414A84A6A63047DDDBAE6D139B7A64565EFF3A8ECA130330603551D1F042C302A3028A026A0248622687474703A2F2F63726C2E726F6F742D78312E6C657473656E63727970742E6F7267307206082B0601050507010104663064303006082B060105050730018624687474703A2F2F6F6373702E726F6F742D78312E6C657473656E63727970742E6F72672F303006082B060105050730028624687474703A2F2F636572742E726F6F742D78312E6C657473656E63727970742E6F72672F301F0603551D2304183016801479B459E67BB6E5E40173800888C81A58F6E99B6E300D06092A864886F70D01010B050003820201006BCCFD71AFF3269F19118C3CFEA19E082823D0C65867F89749B38C186E0A40D5AA81BFEEBB07D72B73B5E78FC62037EE4A92068D1F5EEE4699B58EC34F46A703D8C80018D2258E314EB26EB4C91D2E3AE94AF3A1EFF82302D0A10736D7D10A961FB8520BB5AB82B1358D6A30303FD585F6426F4CE80700B3552F64A1D2A494006EC62B6CDEA86CDA1418E786E8E34762BF0A450CF914C7D961C038F5F99D0D3C0A9FFA722CE90915E03F766FE83FDF5498A085A1E39AE70F5826BD958BB89C5DDE7D094F26A6B5A1C455E799AD685374FC64DA2F8E26F7507A1C8E51097CD6C3CBDCE029361445C315C1EA357646C735B2DCE66223224FE427E44F0FE72147E8098F2DC687DC4C3640EB65029FB083BDDE1C53B1932681943873EBEE8C2D4C8D26A4F808BCED865E6103E1D5F0863124E7A39208237EC3502B489FC049C498B1B8F78E30AB232CB0F902D315CB47E9775F657147A4E1B8C363CFD48DBE7FC1F9F6F8A93BCFBD92C5675E923B473960662B1AAAF230EEB8CC38CA68358399FB47B4D9FD8DF2E1B4E95940E4047168D37C4D336D700450C8AC265979586F0D6E0CB53EBE02CB366DCA681F061C1A897681A2C36B99367641B8D4E1AD5DAD1F63E4E376BC98E1C39CB2D2542A3BF630992EE32693B638C4097D6B26F14B9EA61EC02F859B18E058FBB244DB8B0C058ADF49A6A5C259DC2983E9D16D022E21EC7DFB). 

The Golang x509.Certificate object supports a list of Policy OIDs -- which CFSSL exposes -- but does not appear to permit embedding a sub-sequence matching [RFC 3280 (page 106)'s](https://tools.ietf.org/html/rfc3280.html#page-106) `policyQualifiers`.  The appropriate ASN.1 structure is included below for convenience.

Given that [signer.go's behavior for OCSPNoCheck](https://github.com/cloudflare/cfssl/blob/6bedb78c225f0c7fe813f36e0e26e47eaf5fb142/signer/signer.go#L339) is already to append an `ExtraExtension` to the `x509.Certificate` object, it appears that would be the precedent for  a way forward.

What I would propose is for CFSSL to manually construct the x509 Policies extension off of the Policy OIDs provided in its configuration, and to add a new config structure defining a `cps_url` and a `id-qt-unotice` for a particular `CertPolicyId`. If present, that would trigger the insertion of a policy and qualifier containing the `IA5String` of the URL as well as the string of the User Notice.

```
CertificatePolicies ::= SEQUENCE SIZE (1..MAX) OF PolicyInformation

PolicyInformation ::= SEQUENCE {
     policyIdentifier   CertPolicyId,
     policyQualifiers   SEQUENCE SIZE (1..MAX) OF
             PolicyQualifierInfo OPTIONAL }

CertPolicyId ::= OBJECT IDENTIFIER

PolicyQualifierInfo ::= SEQUENCE {
       policyQualifierId  PolicyQualifierId,
       qualifier        ANY DEFINED BY policyQualifierId }
```

### PR DESCRIPTION
**Pull request by jsha**
Previously we could only add certificate policies, but not their related
qualifiers.

Fixes #230.

### CODE REVIEW

File: config/config.go [hunk 1]
```diff
@@ -59,6 +69,31 @@ type SigningProfile struct {
...
+	parsedOid, err := parseObjectIdentifier(string(data))
+	if err != nil {
+		return err
+	}
+	*oid = OID(parsedOid)
+	log.Debugf("Parsed OID %v", *oid)
+	return
+}
+
+func (oid OID) MarshalJSON() ([]byte, error) {
```
Comment (kisom) [inline config/config.go]: This should probably be replaced with 

```
fmt.Sprintf(`"%v"`, oid)
```

It's a lot shorter and more apparent what's happening, I think.

**Comment by kisom** [comment]
LGTM.

**Comment by kisom** [comment]
Looks like `golint` wasn't happy about a few things:

```
/home/travis/gopath/src/github.com/cloudflare/cfssl/config/config.go:37:6: exported type OID should have comment or be unexported

/home/travis/gopath/src/github.com/cloudflare/cfssl/config/config.go:39:1: comment on exported type CertificatePolicy should be of the form "CertificatePolicy ..." (with optional leading article)

/home/travis/gopath/src/github.com/cloudflare/cfssl/config/config.go:73:1: exported method OID.UnmarshalJSON should have comment or be unexported

/home/travis/gopath/src/github.com/cloudflare/cfssl/config/config.go:87:1: exported method OID.MarshalJSON should have comment or be unexported

/home/travis/gopath/src/github.com/cloudflare/cfssl/signer/signer.go:355:2: struct field PolicyQualifierId should be PolicyQualifierID

/home/travis/gopath/src/github.com/cloudflare/cfssl/signer/signer.go:364:2: comment on exported var IDQTCertificationPracticeStatement should be of the form "IDQTCertificationPracticeStatement ..."

/home/travis/gopath/src/github.com/cloudflare/cfssl/signer/signer.go:368:2: comment on exported var IDQTUserNotice should be of the form "IDQTUserNotice ..."
```

**Comment by jsha** [comment]
Thanks very much for the quick review!

**Comment by kisom** [comment]
And thanks for the feature!

**Pull request merged**